pkg/pub_package_reader/lib/src/pubspec_content_override.dart

```diff
@@ -0,0 +1,170 @@
+import 'package:yaml/yaml.dart' as yaml;
+import 'package:yaml_edit/yaml_edit.dart';
+
+final _logger = Logger('pubspec_content_override');
+
+/// The version regular expression that accepted any character as separator.
+final _lenientRegExp = RegExp(r'(\d+).(\d+).(\d+)' // Version number.
+    r'(-([0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*))?' // Pre-release.
+    r'(\+([0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*))?' // Build.
+    );
+
+/// Packages with bad versions as detected by
+/// https://github.com/dart-lang/pub_semver/pull/63
+const _packagesWithBadVersions = <String>{
+  'assets_audio_player_web',
+  'bloc_pattern',
+  'bokain_models',
+  'built_collection',
+  'built_redux',
+  'built_redux_thunk',
+  'built_value',
+  'built_value_generator',
+  'built_value_test',
+  'built_viewmodel',
+  'built_viewmodel_flutter_built_redux',
+  'built_viewmodel_flutter_redux',
+  'built_viewmodel_generator',
+  'carp_context_package',
+  'codemetrics',
+  'fluttercontactpicker',
+  'interactive_webview',
+  'meta_types',
+  'meta_types_generator',
+  'meta_types_json',
+  'meta_types_json_generator',
+  'meta_types_redux',
+  'meta_types_redux_generator',
+  'mocktail',
+  'mocktail_image_network',
+  'polymer_interop',
+};
+
+/// Override pubspec.yaml content if needed:
+///
+/// - If the archive was created before 2022-01-01, we may need to update the
+///   version number, version constraint and SDK constraints, as `pub_semver`
+///   accepted separator characters other than `.`.
+///   https://github.com/dart-lang/pub_semver/pull/63
+String overridePubspecContentIfNeeded({
+  required String content,
+  required DateTime published,
+}) {
+  // quick checks to skip overrides
+  if (published.year >= 2022) {
+    return content;
+  }
+
+  try {
+    return _fixupBrokenVersionAndConstraints(content);
+  } on FormatException catch (e, st) {
+    _logger.info('Unable to parse pubspec.', e, st);
+  }
+  // fallback: don't override content
+  return content;
+}
+
+/// Fixes broken version and version constraints in [pubspecYaml] and returns a new YAML string.
+///
+/// Prior to [pub_semver#63] version numbers and version constraints were accepted with any
+/// character as separate, instead of requiring  `.` (dot). This occurred because the regular expression
+/// in `package:pub_semver` didn't escape dots as `\.`, but just contained `.` (dot, meaning any character).
+///
+/// Thus, package versions published prior to 2022-01-01, maybe contain _version numbers_,
+/// _version constraints_ and _SDK constraints_ that doesn't use `.` (dot) as separator. As removing
+/// these _package versions_ would break existing users, this function will rewrite the version numbers
+/// and version constraints to have a valid format equivalent to how older versions of `pub_semver`
+/// would have interpreted the version number / version constraints.
+///
+/// [pub_semver#63]: https://github.com/dart-lang/pub_semver/pull/63
+String _fixupBrokenVersionAndConstraints(String pubspecYaml) {
```

Comment:
We do have the `/packages/<pkg>/pubspec` endpoint that displays the original format, and I think it would be nice if these overrides would be still nicely visible there.

Reply:
Should that endpoint ideally not display the original pubspec.yaml (with the errors)?

That would entail storing two pubspecs for each package, maybe that is overkill for such a small exposure...